openml-api/src/main/java/com/feedzai/openml/data/schema/DatasetSchema.java

```diff
@@ -54,14 +55,26 @@
     /**
      * Creates a new instance.
      *
-     * @param targetIndex  The index of the target field.
+     * @param targetIndex  The index of the target field. Use a negative value (or {@link #DatasetSchema(List)}) to represent no
+     *                     target variable.
      * @param fieldSchemas The list of {@link FieldSchema} that represent each feature.
      */
     public DatasetSchema(final int targetIndex,
```

Comment:
+1

Reply:
furthermore, should we re-arranje the type hierarchy to account for Unsupervised Models? How do we distinguish between supervised and unsupervised models? 

would it be beneficial to have a DatasetSchema specific implementation for supervised learning and one for unsupervised learning, each parameterised in  the corresponding machine learning model hierarchy?

Reply:
Regarding the fields, I think it's up for discussion, and a bit out of context for this task (I can't use Optional in the fields because it is not serializable).

Regarding what @pedrorijo91 said, I'm feel free to correct if I'm wrong. I had already considered that option:
 - Have a `DatasetSchema` which would be a common type between schemas for supervised and unsupervised models.
 - Have a `SupervisedSchema` (name up for debate), which would extend `DatasetSchema` with a `int#getTargetIndex`

(not really sure where the parameterization would get here)

If this was also your suggestion @pedrorijo91, then the downside I see is that each time we want to get the target index, we have to downcast according to the algorithm type (will answer your question below), which is a counter-intuitive with the polymorphic approach. Instead, I followed this road as in fact (disregarding machine learning models), a target variable should be a required characteristic of a schema, as schemas with no target variable are in a sense the "default" concept of a schema.

Regarding:
> How do we distinguish between supervised and unsupervised models?

It will be a separate pull request, but >spoiler alert< we will add a new enum value to `MachineLearningAlgorithmType`, where the gist of it is that we do not separate by supervised/unsupervised, but instead by Binary Classification/Anomaly Detection/Multi Classification/etc.

Reply:
Ah, that serialization implementation, I had missed that detail: but then is that documented in the implementation comments of the class? I'm ok with keeping the int as long as that is documented :)

I agree with your comment for Rijo: I think for now (in this PR) I'd keep it as is, and as you change the rest of the code and use it, try to get the feeling if this is really needed.

Reply:
As discussed offline, this seems to be the best approach given our requirements